ext/MitsubaLoader/CMitsubaLoader.h

```diff
@@ -16,6 +16,304 @@ namespace ext
+		float dummy2;
+		//if flag decides to use diffuse refl texture, diffReflectance[0..1] is bindless texture ID
+		//otherwise diffReflectance[0..2] is constant diffuse refl
+		float diffReflectance[3];
+	} PACK_STRUCT;
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		float thickness;
+		vec2 int_ext_ior;
+		vec2 alpha_u;
+		vec2 alpha_v;
+		vec3 specReflectance;
+		vec3 specTransmittance;
+		vec3 sigmaA;
+	};
+	*/
+	struct alignas(16) SAllCoating
+	{
+		//child index (into BSDF buffer) is stored on 15 highest bits
+		uint32_t bitfields;
+		float thickness;
+		//same as SAllDielectric::int_ext_ior
+		float int_ext_ior[2];
+		//same as for SAllDielectric::alpha_u,alpha_v
+		float alpha_u[2];
+		float alpha_v[2];
+		//same as SAllDielectric::specReflectance
+		float specReflectance[3];
+		float dummy0;
+		//same as SAllDielectric::specTransmittance
+		float specTransmittance[3];
+		float dummy1;
+		float sigmaA[3];
+	} PACK_STRUCT;
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		vec2 exponent;
+		vec3 specReflectance;
+		vec3 diffReflectance;
+	};
+	*/
+	struct alignas(16) SPhong
+	{
+		uint32_t bitfields;
+		float dummy0;
+		//if flag decides to use exponent texture, exponent[0..1] is bindless texture ID
+		//otherwise exponent[0] is constant exponent
+		float exponent[2];
+		//same as SAllDielectric::specReflectance
+		float specReflectance[3];
+		float dummy1;
+		//same as SAllPlastic::diffReflectance
+		float diffReflectance[3];
+	} PACK_STRUCT;
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		vec2 alpha_u;
+		vec2 alpha_v;
+		vec3 specReflectance;
+		vec3 diffReflectance;
+	};
+	*/
+	struct alignas(16) SWard
+	{
+		uint32_t bitfields;
+		float dummy0;
+		//same as for SAllDielectric::alpha_u,alpha_v
+		float alpha_u[2];
+		float alpha_v[2];
+		//same as SAllDielectric::specReflectance
+		float specReflectance[3];
+		float dummy1;
+		//same as SAllPlastic::diffReflectance
+		float diffReflectance[3];
+	} PACK_STRUCT;
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		uint weightsOffset;
+	};
+	*/
+	struct alignas(16) SMixture
+	{
+		//children index (into BSDF buffer) is stored on 15 highest bits
+		uint32_t bitfields;
+		//index into weights buffer
+		uint32_t weightsOffset;
+	} PACK_STRUCT;
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		vec2 weight;
+	};
+	*/
+	struct alignas(16) SBlend
+	{
+		//children index (into BSDF buffer) is stored on 15 highest bits
+		uint32_t bitfields;
+		float dummy0;
+		//if flag decides to use weight texture, weight[0..1] is bindless texture ID
+		//otherwise weight[0] is constant blend weight
+		float weight[2];
+	} PACK_STRUCT;
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		vec3 opacity;
+	};
+	*/
+	struct alignas(16) SMask
+	{
+		//child index (into BSDF buffer) is stored on 15 highest bits
+		uint32_t bitfields;
+		float dummy0[3];
+		//if flag decides to use opacity texture, opacity[0..1] is bindless texture ID
+		//otherwise opacity[0..2] is constant per-channel opacity
+		float opacity[3];
+	} PACK_STRUCT;
```

Comment:
upgrade the SBlend BSDF to have a 3-channel weight (packed into RGB19E7 - our custom format for replacing vec3).

then you can decompose the Mask BSDF as a (textured) blend BSDF between the "fully transparent BSDF" (delta distribution centered around the transmission direction [negative of viewer direction], divided by absolute value of the cosine between the normal and outgoing direction) and the child of the original mask BSDF

Reply:
>then you can decompose the Mask BSDF as a (textured) blend BSDF between the "fully transparent BSDF (...)"

not really because mask BSDF gives per-channel opacity (idk what that means but.. yea XD)

Reply:
well, upgrade the "blend" BSDF to do a 3-channel blend ;)

Reply:
i.e. if it was an actual `BlendBSDF` then use a RRRA swizzle on the sampler, if it was a `MaskBSDF` then use a RGBA swizzle